src/main/java/eu/europeana/metis/sandbox/service/dataset/DatasetReportServiceImpl.java

```diff
@@ -41,197 +43,205 @@
+            RecordLogRepository recordLogRepository,
+            RecordErrorLogRepository errorLogRepository) {
+        this.datasetRepository = datasetRepository;
+        this.recordLogRepository = recordLogRepository;
+        this.errorLogRepository = errorLogRepository;
     }
 
-    if (stepStatistics.isEmpty()) {
-      return new ProgressInfoDto(getPublishPortalUrl(dataset, 0L, 0L),
-          dataset.getRecordsQuantity(), 0L, List.of(),
-          datasetInfoDto);
+    @Override
+    @Transactional(readOnly = true)
+    public ProgressInfoDto getReport(String datasetId) {
+        requireNonNull(datasetId, "Dataset id must not be null");
+
+        // search for dataset
+        DatasetEntity dataset = getDataset(datasetId);
+
+        //Create DatasetInfoDto from DatasetEntity
+        DatasetInfoDto datasetInfoDto = new DatasetInfoDto(datasetId, dataset.getDatasetName(), dataset.getCreatedDate(),
+                dataset.getLanguage(), dataset.getCountry(), dataset.getRecordLimitExceeded(),
+                StringUtils.isNotBlank(dataset.getXsltEdmExternalContent()));
+
+        // pull records and errors data for the dataset
+        List<StepStatistic> stepStatistics;
+        List<ErrorLogView> errorsLog;
+        try {
+            stepStatistics = recordLogRepository.getStepStatistics(datasetId);
+            errorsLog = errorLogRepository.getByRecordIdDatasetId(datasetId);
+        } catch (RuntimeException exception) {
+            throw new ServiceException(format("Failed to get report for dataset id: [%s]. ", datasetId),
+                    exception);
+        }
+
+        if (stepStatistics.isEmpty()) {
+            return new ProgressInfoDto(getPublishPortalUrl(dataset, 0L),
+                    dataset.getRecordsQuantity(), 0L, List.of(),
+                    datasetInfoDto, getErrorMessage(0L, 0L, dataset));
+        }
+
+        // get qty of records completely processed
+        long completedRecords = getCompletedRecords(stepStatistics);
+
+        // get qty of records that failed
+        long failedRecords = getFailedRecords(stepStatistics);
+
+        // get records processed by step
+        Map<Step, Map<Status, Long>> recordsProcessedByStep = getStatisticsByStep(
+                stepStatistics);
+
+        // get errors by step
+        Map<Step, Map<Status, Map<String, List<ErrorLogView>>>> recordErrorsByStep = getRecordErrorsByStep(
+                errorsLog);
+
+        // collect steps processing information
+        List<ProgressByStepDto> stepsInfo = new LinkedList<>();
+        recordsProcessedByStep.forEach((step, statusMap) -> addStepInfo(stepsInfo, statusMap, step,
+                recordErrorsByStep));
+
+        return new ProgressInfoDto(
+                getPublishPortalUrl(dataset, completedRecords),
+                dataset.getRecordsQuantity(), completedRecords,
+                stepsInfo, datasetInfoDto, getErrorMessage(completedRecords, failedRecords, dataset));
```

Comment:
my suggestion is: ```getErrorMessage(completedRecords, failedRecords, dataset.getRecordsQuantity())``` to keep variables at the same level.

Reply:
Done